typing/parmatch.ml

```diff
@@ -2167,11 +2342,26 @@ let rec matrix_stable_vars rs = match rs with
     (* The stable variables are those stable at any position *)
     List.fold_left IdSet.union IdSet.empty stables_in_varsets
 | rs ->
-   let submatrices = split_rows (simplify_first_col rs) in
-   let submat_stable = List.map matrix_stable_vars submatrices in
-   (* a stable variable must be stable in each submatrix;
-      if the matrix has at least one row, there is at least one submatrix *)
-   reduce IdSet.inter submat_stable
+    let rs = simplify_first_col rs in
+    if not (first_column_is_coherent rs) then (
+      (* If the first column is incoherent, then all the variables of this
+         matrix are stable. *)
+      List.fold_left (fun acc (_, { varsets; _ }) ->
+        List.fold_left IdSet.union acc varsets
+      ) IdSet.empty rs
```

Comment:
I have worked on a better implementation of this on my own (a distinguish `All` value), as I thought that this would be outside the scope of your PR. I hope to send a PR for this soon.

Reply:
This is now pushed as https://github.com/ocaml/ocaml/pull/1552/commits/63ed2cc031a49aaec3c0ad7b64cd2a44879e3d6a, please take a look, feel free to ask for changes, and eventually to integrate into the present PR if you wish to -- I can rebase my own on top of the final version of this commit.

Reply:
The idea of your implementation seems nice indeed (I haven't actually reviewed the actual code just yet), but IMO they are independent (which is a good thing) and should be reviewed and merged separately (that is, I don't want to "merge" the GPRs together before actually merging into trunk).